local-cluster/tests/local_cluster.rs

```diff
@@ -1804,17 +1809,30 @@ fn do_test_optimistic_confirmation_violation_with_or_without_tower(with_tower: b
 
     // monitor for actual votes from validator A
     let mut bad_vote_detected = false;
+    let mut a_votes = vec![];
     for _ in 0..100 {
         sleep(Duration::from_millis(100));
 
         if let Some(last_vote) = last_vote_in_tower(&val_a_ledger_path, &validator_a_pubkey) {
-            if votes_on_c_fork.contains(&last_vote) {
-                bad_vote_detected = true;
-                break;
+            a_votes.push(last_vote);
+            let blockstore = Blockstore::open_with_access_type(
+                &val_a_ledger_path,
+                AccessType::TryPrimaryThenSecondary,
+                None,
+            )
+            .unwrap();
+            let ancestors = AncestorIterator::new(last_vote, &blockstore);
+            for a in ancestors {
+                if votes_on_c_fork.contains(&a) {
+                    bad_vote_detected = true;
+                    break;
```

Comment:
well, this only breaks from the newly-added inner `for` loop, maybe `break 'outer` or use `if ancestors.any(|a| votes_on_c_fork.contains(a)) { bad_vote_detected = true; break}`.

I'd prefer the latter. nested loops is hard to manage like this bug. ;)